File: RemoveDuplicatesFromSortedArray.cpp
// Time Complexity :  O(n) where n = input size. 
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes

// Three line explanation of solution in plain english

/* Use two pointers i, j and count variable to maintain the occurrence count of current element. 
 * Iterate over array using pointer i, if (i)th element and (i-1)th element are same then increment the count.
 * If count is less than three then replace the (j)th element of i/p array with (i)th element and increment j.
 * At the end return j.
 */

class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        if (nums.size() == 0)
            return 0;
        
        int j = 1, count = 1;
        
        for (int i = 1; i < nums.size(); i++)
        {
            if (nums[i] == nums[i - 1])
            {
                count++;
            }
            else
            {
                count = 1;
            }
            
            if (count < 3)
            {
                nums[j] = nums[i];
                j++;
            }
        }
        return j;
    }
};

File: RemoveDuplicatesFromSortedArray.js
// Time Complexity : O (n)
// Space Complexity : O (1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english : Using two pointers i, j and traverse through the nums array. Replace the jth index with ith index if count <=2 and return the j that holds the sorted array with at max two repeated values.

/**
 * @param {number[]} nums
 * @return {number}
 */
var removeDuplicates = function(nums) {
  if (nums.length === 0 || nums === null) return 0;

  var n = nums.length;
  var j = 1;
  var count = 1;

  for (var i = 1; i < n; i++) {
    if (nums[i] === nums[i - 1]) {
      count++;
    } else {
      count = 1;
    }
    if (count <= 2) {
      nums[j] = nums[i];
      j++;
    }
  }
  return j;
};
